independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java

```diff
@@ -0,0 +1,146 @@
+                    // we stop if there's no more to append
+                    if (len == 0) {
+                        return toWrite;
+                    }
+                    off += toWrite;
+                    alreadyWritten = toWrite;
+                }
+            }
+        }
+        final int availableCapacity = capacity - size;
+        if (availableCapacity == 0) {
+            return alreadyWritten;
+        }
+        // we can still write some
+        int toWrite = Math.min(len, availableCapacity);
+        final int chunkCapacity;
+        if (minChunkSize > 0) {
+            // Cannot allocate less the minChunkSize, till the limit of capacity left
+            chunkCapacity = Math.min(Math.max(minChunkSize, toWrite), availableCapacity);
+        } else {
+            chunkCapacity = toWrite;
+        }
+        var tmpBuf = allocator.directBuffer(chunkCapacity);
+        tmpBuf.writeBytes(bytes, off, toWrite);
+        if (buffer == null) {
+            buffer = tmpBuf;
+        } else {
+            try {
+                if (otherBuffers == null) {
+                    otherBuffers = new ArrayDeque<>();
+                }
+                otherBuffers.add(tmpBuf);
+            } catch (Throwable t) {
+                size -= alreadyWritten;
+                tmpBuf.release();
+                throw t;
+            }
+        }
+        size += toWrite;
+        return toWrite + alreadyWritten;
+    }
+
+    public ByteBuf flushBatch() {
+        var firstBuf = buffer;
+        if (firstBuf == null) {
+            return null;
+        }
+        var others = otherBuffers;
+        if (others == null || others.isEmpty()) {
+            size = 0;
+            buffer = null;
+            // super fast-path
+            return firstBuf;
+        }
+        var batch = allocator.compositeDirectBuffer(1 + others.size());
+        try {
+            // we unset this first in case addComponent would fail and will be released
+            // same applies for size
+            buffer = null;
+            final int firstBytes = firstBuf.readableBytes();
+            size -= firstBytes;
+            batch.addComponent(true, 0, firstBuf);
+            for (int i = 0, othersCount = others.size(); i < othersCount; i++) {
+                var curr = others.poll();
+                // we have already been able to remove it from the queue!
+                size -= curr.readableBytes();
+                // if addComponent fail, it takes care of releasing curr and throwing the exception:
+                // we lose curr data, but not the rest into others
+                batch.addComponent(true, 1 + i, curr);
+            }
+            assert size == 0;
+            // net-safe barrier to keep this to work in case of bugs
+            size = 0;
+            return batch;
+        } catch (Throwable anyError) {
+            // we cannot reuse what's fallen already into batch :"(
+            batch.release();
```

Comment:
Probably worth dropped the whole existing content (into the queue) here: it's an all-or-nothing use case